### PR TITLE
Convert to ts-tcpip and use ts-simactuators

### PR DESCRIPTION
**Pull request by couger01**

### CODE REVIEW

File: python/lsst/ts/athexapod/controller.py [hunk 1]
```diff
@@ -80,36 +73,30 @@ def is_connected(self):
...
-                )
-
-            connect_task = asyncio.open_connection(host=self.host, port=self.port)
-
-            self.reader, self.writer = await asyncio.wait_for(
-                connect_task, timeout=self.long_timeout
-            )
+        self.client = tcpip.Client(
+            host=self.host,
+            port=self.port,
```
Comment (wvreeven) [inline python/lsst/ts/athexapod/controller.py:82]: Where are `self.host` and `self.port` defined? They were removed from the `__init__` function.

File: python/lsst/ts/athexapod/controller.py [hunk 2]
```diff
@@ -55,20 +57,11 @@ class ATHexapodController:
     """
 
     def __init__(self, log=None):
-        self.host = "127.0.0.1"
-        self.port = 50000
-        self.timeout = 2.0
-        self.long_timeout = 30.0
```
Comment (wvreeven) [inline python/lsst/ts/athexapod/controller.py:61]: Please update the class doc string and remove these attributes.

File: python/lsst/ts/athexapod/controller.py [hunk 3]
```diff
@@ -129,33 +116,27 @@ async def write_command(self, cmd, has_response=True, num_line=1):
...
-                except asyncio.TimeoutError:
-                    self.log.warning(
-                        "Timed out waiting for response from controller. Result "
-                        "may be incomplete."
-                    )
-                    raise
-
-                return replies
+        if self.client.connected:
+            async with self.lock:
```
Comment (wvreeven) [inline python/lsst/ts/athexapod/controller.py:122]: You have removed the use of this lock everywhere else so remove it here as well. And remove it from `__init__`.

File: python/lsst/ts/athexapod/controller.py [hunk 4]
```diff
@@ -129,33 +116,27 @@ async def write_command(self, cmd, has_response=True, num_line=1):
...
+                await self.client.write_str(cmd)
+                if has_response:
+                    try:
+                        replies = []
+                        for _ in range(num_line):
```
Comment (wvreeven) [inline python/lsst/ts/athexapod/controller.py]: Please rename `num_line` to `_num_lines`.

File: python/lsst/ts/athexapod/controller.py [hunk 5]
```diff
@@ -247,8 +228,8 @@ async def controller_ready(self):
         """
         ret = await self.write_command("\7", num_line=1)
 
-        comp = ret == chr(177)
-        self.log.debug(f"ret={ret} : {chr(177)} : comp={comp}")
+        comp = ret[0] == chr(177)
```
Comment (wvreeven) [inline python/lsst/ts/athexapod/controller.py]: Please extract `chr(177)` to a constant with a meaningful name and use that in this line and the next.

File: towncrier.toml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+[tool.towncrier]
```
Comment (wvreeven) [inline towncrier.toml]: Why is this file checked in? Is it different than the one created by ts_pre_commit_conf?

File: ups/ts_athexapod.table [hunk 7]
```diff
@@ -2,9 +2,6 @@
 # - Any package whose API is used directly should be listed explicitly.
 # - Common third-party packages can be assumed to be recursively included by
 #   the "sconsUtils" package.
-setupRequired(ts_salobj)
-setupRequired(ts_idl)
-setupRequired(ts_config_attcs)
```
Comment (wvreeven) [inline ups/ts_athexapod.table]: Why did you remove these lines? ATHexapod still depends on those packages, right?

File: python/lsst/ts/athexapod/csc.py [hunk 8]
```diff
@@ -42,6 +41,10 @@
 REFERENCING_ERROR = 104
 
 
+def execute_csc():
```
Comment (wvreeven) [inline python/lsst/ts/athexapod/csc.py]: I think we decided to name this `run_csc` so please rename.

File: python/lsst/ts/athexapod/csc.py [hunk 9]
```diff
@@ -186,129 +176,53 @@ async def configure(self, config):
...
-                await asyncio.wait_for(
-                    self.wait_movement_done(), timeout=self.config.reference_timeout
-                )
-            except asyncio.TimeoutError as e:
-                self.log.error("Referencing time out.")
-                self.log.exception(e)
-                await self.fault(
-                    code=REFERENCING_TIMEOUT,
-                    report="Referencing time out.",
-                    traceback=traceback.format_exc(),
```
Comment (wvreeven) [inline python/lsst/ts/athexapod/csc.py]: I am not happy with any of these changes. This class was designed like that with a reason and changing this without further explanation will not do. Please revert.

**Review by wvreeven — CHANGES_REQUESTED** [review]
There are several design decisions that I disagree with as stated in my comments. On top of that, you have removed attributes from the `ATHexapodController` class that should not have been removed. Please review my comments. When done, request Tiago's review since he knows this project much better than me.

File: towncrier.toml (hunk 6 above)
Comment (couger01) [reply towncrier.toml]: Yeah it is.

File: python/lsst/ts/athexapod/controller.py (hunk 3 above)
Comment (couger01) [reply python/lsst/ts/athexapod/controller.py:122]: I have the lock here since this is the only method that read and writes to the controller.

File: python/lsst/ts/athexapod/controller.py (hunk 1 above)
Comment (couger01) [reply python/lsst/ts/athexapod/controller.py:82]: They are defined in the configure method but I added them back in the __init__ since that is confusing and prone to bugs.

File: python/lsst/ts/athexapod/controller.py (hunk 3 above)
Comment (wvreeven) [reply python/lsst/ts/athexapod/controller.py:122]: But it literally is the only line that uses the lock so it is redundant and you should remove it.

File: python/lsst/ts/athexapod/controller.py (hunk 1 above)
Comment (wvreeven) [reply python/lsst/ts/athexapod/controller.py:82]: Setting them in the config method is not enough. All attributes need to be defined in the `__init__` method. Thanks for adding them back.

**Review by wvreeven — COMMENTED** [review]
Thanks for making some of the changes I requested. I still don't understand why that lock is needed and I really think it should be removed.

**Review by wvreeven — COMMENTED** [review]
Thanks for the additional changes.

File: python/lsst/ts/athexapod/csc.py [hunk 10]
```diff
@@ -154,18 +157,6 @@ async def configure(self, config):
         config
         """
         host = config.host
-        if host.startswith("ENV."):
```
Comment (tribeiro) [inline python/lsst/ts/athexapod/csc.py:157]: Why did you removed this? this is a feature required to run the system at the test stands.

File: python/lsst/ts/athexapod/csc.py [hunk 11]
```diff
@@ -383,27 +382,23 @@ async def do_moveToPosition(self, data):
         )
 
         try:
+            await self.evt_positionUpdate.set_write(
```
Comment (tribeiro) [inline python/lsst/ts/athexapod/csc.py]: I think the previous code was the correct behavior... You first wait for the move to complete , than you update the position, not the other way around.

File: python/lsst/ts/athexapod/csc.py [hunk 12]
```diff
@@ -383,27 +382,23 @@ async def do_moveToPosition(self, data):
...
+            )
             await asyncio.wait_for(
                 self.wait_movement_done(), timeout=self.config.movement_timeout
             )
+            await self.evt_inPosition.set_write(inPosition=True, force_output=True)
+            await self.report_detailed_state(ATHexapod.DetailedState.NOTINMOTION)
         except Exception as e:
             self.log.exception("Error executing moveToPosition command")
             raise e
-        else:
```
Comment (tribeiro) [inline python/lsst/ts/athexapod/csc.py:392]: I think removing the else here and moving the in position above is ok, but the finally statement needs to be untouched. I would just revert these changes altogether because they are not doing any good.

File: python/lsst/ts/athexapod/mock_server.py
Comment (tribeiro) [inline python/lsst/ts/athexapod/mock_server.py:1]: you included the "w" parameter in some methods that didn't had it before. I am not sure about this change.

**Review by tribeiro — CHANGES_REQUESTED** [review]
Here my initial round of review.


To be honest, I am not sure about these changes. I think going to the tcpip client and using the simactuators looks like a nice improvement. However it is hard to judge the changes since they are conflated with other completely unrelated changes. Some of them actually remove functionality we need like the ENV bit.

Please, consider cleaning up the changes to only do the tcp/ip and simactutors changes.

File: python/lsst/ts/athexapod/mock_server.py
Comment (couger01) [reply python/lsst/ts/athexapod/mock_server.py:1]: Where?

File: python/lsst/ts/athexapod/controller.py [hunk 13]
```diff
@@ -80,38 +72,32 @@ def is_connected(self):
...
-                    await asyncio.wait_for(self.writer.drain(), timeout=self.timeout)
-                finally:
-                    self.writer.close()
-                    self.writer = None
+        try:
+            await self.client.close()
+        except Exception:
+            self.log.exception("Failed to disconnect")
+        finally:
+            self.client = tcpip.Client(host="", port=None, log=self.log)
```
Comment (tribeiro) [inline python/lsst/ts/athexapod/controller.py:98]: Uhm.. why are you recreating the client here? shouldn't this be set to `None`?

File: python/lsst/ts/athexapod/controller.py [hunk 14]
```diff
@@ -129,33 +115,30 @@ async def write_command(self, cmd, has_response=True, num_line=1):
...
+                        self.log.debug(f"{replies=}")
+                        return replies
+                    except asyncio.TimeoutError:
+                        self.log.warning(
+                            "Timed out waiting for response from controller. Result "
+                            "may be incomplete."
+                        )
+                        raise
+        else:
+            raise RuntimeError("Not connected.")
```
Comment (tribeiro) [inline python/lsst/ts/athexapod/controller.py]: Please, add a more detailed error message.

File: python/lsst/ts/athexapod/csc.py [hunk 15]
```diff
@@ -154,18 +158,18 @@ async def configure(self, config):
...
         host = config.host
+        # Support teststand connection via Environment Variable.
         if host.startswith("ENV."):
             env_name = host.split("ENV.")[1]
             if env_name not in os.environ:
-                raise RuntimeError(
-                    "Failed to configure CSC."
-                    f"Environment variable {env_name} is not defined."
-                )
+                msg = f"""Failed to configure CSC
```
Comment (tribeiro) [inline python/lsst/ts/athexapod/csc.py]: This seems like a pretty innocuous change. Please, revert back to the original code.

File: python/lsst/ts/athexapod/csc.py [hunk 16]
```diff
@@ -154,18 +158,18 @@ async def configure(self, config):
...
+                Environment variable {env_name} is not defined.
+                """
+                raise RuntimeError(msg)
             host = os.environ[env_name]
             self.log.debug(
-                f"Reading hexapod controller host from environment variable. {env_name}={host}."
```
Comment (tribeiro) [inline python/lsst/ts/athexapod/csc.py]: Also pretty innocuous change. Please, revert back.

File: python/lsst/ts/athexapod/csc.py [hunk 17]
```diff
@@ -481,38 +489,36 @@ async def telemetry(self):
...
+                    await self.fault(code=error, report=translate_error(error))
+                else:
+                    await asyncio.sleep(self.heartbeat_interval / sub_tasks)
+            except asyncio.CancelledError:
+                self.log.info("Telemetry task cancelled.")
+            except RuntimeError:
+                self.log.exception("ATHexapod unexpectedly disconnected.")
                 await self.fault(
-                    code=error, report=translate_error(error), traceback=""
+                    code=100000, report="ATHexapod unexpectedly disconnected."
```
Comment (tribeiro) [inline python/lsst/ts/athexapod/csc.py:524]: Why are you changing this?

**Review by tribeiro — COMMENTED** [review]
I still have some concerns over this PR.

One thing I just realized is that you introduced this `parameterized` package. I think we discussed adding this in the past and decided against using it. If you really want to add it, we should discuss with the team and probably add it to ts-conda-build. Please, don't include isolated dependencies 

Also, I was looking over the changes you made to support simulation mode. Although I agree this is probably useful, the reason for not supporting simulation mode on this CSC is because we have a vendor provided simulator we use on TTS and BTS. I am a bit on the fence about this change. I am ok letting it past if you add some documentation in the class about it. 

Also, please, see my inline comments.

File: python/lsst/ts/athexapod/controller.py (hunk 13 above)
Comment (couger01) [reply python/lsst/ts/athexapod/controller.py:98]: The way that the control flow works is that the client is initialized with an empty host and then recreate with the correct configuration and then recreated to empty host because I want to know whether the client is connected at any time as opposed to using None.

File: python/lsst/ts/athexapod/csc.py (hunk 17 above)
Comment (couger01) [reply python/lsst/ts/athexapod/csc.py:524]: I'm adding that fault transition. It's for if the ATHexapod is unexpectedly disconnected. The error fault transition is in 511-512
